junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/TestTemplateTestDescriptor.java

```diff
@@ -142,13 +146,16 @@ private void execute(DynamicTestExecutor dynamicTestExecutor, TestDescriptor tes
 		dynamicTestExecutor.execute(testDescriptor);
 	}
 
-	private void validateWasAtLeastInvokedOnce(int invocationIndex,
+	private void validateWasAtLeastInvokedOnce(int invocationIndex, ExtensionContext extensionContext,
 			List<TestTemplateInvocationContextProvider> providers) {
 
-		Preconditions.condition(invocationIndex > 0,
+		boolean allMayReturnEmptyContext = providers.stream().allMatch(
+			p -> p.mayReturnEmptyInvocationContext(extensionContext));
+
+		Preconditions.condition(invocationIndex > 0 || allMayReturnEmptyContext,
 			() -> "None of the supporting " + TestTemplateInvocationContextProvider.class.getSimpleName() + "s "
-					+ providers.stream().map(provider -> provider.getClass().getSimpleName()).collect(
-						joining(", ", "[", "]"))
+					+ providers.stream().filter(p -> !p.mayReturnEmptyInvocationContext(extensionContext)).map(
```

Comment:
I don't think we should filter them out here.

Reply:
@marcphilipp I added this filter to narrow down the list. There is no need for a developer to investigate well-behaving providers, so their presence on the error message would be useless.
Why do you think we should keep them?

Reply:
I agree. Per-provider validation will be most precise. I've added one more commit to the pull request.
Just a side note: this will be a more strict check in case of multiple providers. But sounds like a corner case to me.